### ISSUE TITLE
Creating SNS subscription with SQS Protocol

### ISSUE THREAD

**Issue by rj425** [aws-controllers-k8s/community#1491]
I have a SNS topic and SQS queue. I would like to have my SQS queue subscribed to SNS topic. I have looked at the list of CRDs that SQS and SNS controller offer. But i am unable to find the right CRD that can help me do so.

What have i tried so far:
- I created a subscription object from the AWS console to understand the configuration changes that takes place in the policy. I tried to make these exact changes to SQS policy using SQS controller. The policy gets changed successfully but there is no SNS subscription.
- I have looked at the PlatformApplication and PlatformEndpoint CRDs in SNS controller, but i fail to understand if these two CRDs are the solution. If that is the case, can you point me to the documentation that explains it well.

Thanks you.

**Comment by A-Hilaly** [comment]
Currently we are not able to regenerate the sns-controller/sqs-controller and any new resource because of the problems mentioned in https://github.com/aws-controllers-k8s/community/issues/1292

**Comment by jessebye** [comment]
Hi @A-Hilaly, issue #1292 is now closed. Would you be able to regenerate the SNS controller resources at this point? We need to create SNS subscriptions as well, and are blocked until that is added.

**Comment by A-Hilaly** [comment]
Hi @jessebye the sns-controller have been regenerated and a new release `v0.0.4` was released? currently it doesn't support Subscriptions but we can definitely open an issue and give it a shot.

**Comment by A-Hilaly** [comment]
@jessebye I just checked the API and i don't see a `CreateSubscription` API Call, however there is a `Subscribe` call that is related to SNS Endpoint and SNS Topics, is this the operation you'd like to see supported?

**Comment by jessebye** [comment]
Hi @A-Hilaly , thanks for the update! Yes, `Subscribe` would be the operation we want to support. Also probably `ConfirmSubscription` so we can confirm the subscription after creating it. Thank you 😀

**Comment by jaypipes** [comment]
> Hi @A-Hilaly , thanks for the update! Yes, `Subscribe` would be the operation we want to support. Also probably `ConfirmSubscription` so we can confirm the subscription after creating it. Thank you grinning

@jessebye @A-Hilaly I suppose we could try to create a Subscription resource and have `Subscribe` be the Create call and `Unsubscribe` be the Delete call. We would use the GetSubscriptionAttributes and SetSubscriptionAttributes calls for the Update code path.

This would come with a caveat, though:

We would only support creating Subscription resources to endpoints in the same AWS account. The reason for this is because the requirement for the *endpoint owner* to call [ConfirmSubscription](https://docs.aws.amazon.com/sns/latest/api/API_ConfirmSubscription.html) when receiving a Subscribe call from an account other than the endpoint's owner is antithetical to the way that Kubernetes controllers work to reconcile the state of an owned resource. In the case of Subscribing to a different account's Endpoint, the subscription isn't actually owned by the calling account.

@jessebye would you be cool with the above behaviour?

**Comment by jessebye** [comment]
@jaypipes sounds reasonable to me! Thanks!

**Comment by jessebye** [comment]
@jaypipes @A-Hilaly Is there anything I can do to help move this along? This is blocking our team's progress on some initiatives that use SNS. I'd be glad to contribute if that would be helpful!

**Comment by jaypipes** [comment]
> @jaypipes @A-Hilaly Is there anything I can do to help move this along? This is blocking our team's progress on some initiatives that use SNS. I'd be glad to contribute if that would be helpful!

Working on this right now, @jessebye! :) Just got done [adding FIFOTopic and other new attributes](https://github.com/aws-controllers-k8s/sns-controller/pull/17) to the controller. Adding support for Subscription resource now...

**Comment by matheushent** [comment]
> #1292

Hi Jay, it's been a little bit more than a month and I know you're pretty much busy implementing all the stuff we ask, but, can you give an update on the implementation status?

**Comment by jaypipes** [comment]
@matheushent I've made great progress on it. Should have something pushed by end of today.

**Comment by matheushent** [comment]
@jaypipes awesome mate, do you have the PR or branch so I can follow?

**Comment by jaypipes** [comment]
@matheushent yup! here it is: https://github.com/jaypipes/ack-sns-controller/tree/subscription

I'm working on e2e tests and debugging failures :)

**Comment by jaypipes** [comment]
@matheushent @jessebye @rj425 OK all, I have support for SNS Subscriptions in a PR here:

https://github.com/aws-controllers-k8s/sns-controller/pull/26

Hoping to get that merged today and work on support for changing Subscription Attributes and setting tags on Subscription resources.

**Comment by logand22** [comment]
Would something like `EndpointRef` be possible with the Subscription resource? Would maybe be a QoL improvement when working with Topic resource and SQS Queue resource.

**Comment by jaypipes** [comment]
> Would something like `EndpointRef` be possible with the Subscription resource? Would maybe be a QoL improvement when working with Topic resource and SQS Queue resource.

@logand22 which specific field on Subscription resources do you want to add resource reference ability?
